mat64/gsvd.go

```diff
@@ -0,0 +1,296 @@
+// If the input slice is non-nil, the values will be stored in-place into the slice.
+// In this case, the slice must have length min(m,n)-k, and GeneralizedValues will
+// panic with matrix.ErrSliceLengthMismatch otherwise. If the input slice is nil,
+// a new slice of the appropriate length will be allocated and returned.
+//
+// GeneralizedValues will panic if the receiver does not contain a successful factorization.
+func (gsvd *GSVD) GeneralizedValues(v []float64) []float64 {
+	if gsvd.kind == 0 {
+		panic("gsvd: no decomposition computed")
+	}
+	if v == nil {
+		v = make([]float64, min(gsvd.m, gsvd.n)-gsvd.k)
+	}
+	if len(v) != min(gsvd.m, gsvd.n)-gsvd.k {
+		panic(matrix.ErrSliceLengthMismatch)
+	}
+	floats.DivTo(v, gsvd.c[gsvd.k:min(gsvd.m, gsvd.n)], gsvd.s[gsvd.k:min(gsvd.m, gsvd.n)])
+	return v
+}
+
+// Values1 returns the singular values of the factorized matrix.
+// If the input slice is non-nil, the values will be stored in-place into the slice.
+// In this case, the slice must have length min(m,n)-k, and Values1 will panic with
+// matrix.ErrSliceLengthMismatch otherwise. If the input slice is nil,
+// a new slice of the appropriate length will be allocated and returned.
+//
+// Values1 will panic if the receiver does not contain a successful factorization.
+func (gsvd *GSVD) Values1(c []float64) []float64 {
+	if gsvd.kind == 0 {
+		panic("gsvd: no decomposition computed")
+	}
+	if c == nil {
+		c = make([]float64, min(gsvd.m, gsvd.n)-gsvd.k)
+	}
+	if len(c) != min(gsvd.m, gsvd.n)-gsvd.k {
+		panic(matrix.ErrSliceLengthMismatch)
+	}
+	copy(c, gsvd.c[gsvd.k:min(gsvd.m, gsvd.n)])
+	return c
+}
+
+// Values2 returns the singular values of the factorized matrix.
+// If the input slice is non-nil, the values will be stored in-place into the slice.
+// In this case, the slice must have length min(m,n)-k, and Values2 will panic with
+// matrix.ErrSliceLengthMismatch otherwise. If the input slice is nil,
+// a new slice of the appropriate length will be allocated and returned.
+//
+// Values2 will panic if the receiver does not contain a successful factorization.
+func (gsvd *GSVD) Values2(s []float64) []float64 {
+	if gsvd.kind == 0 {
+		panic("gsvd: no decomposition computed")
+	}
+	if s == nil {
+		s = make([]float64, min(gsvd.m, gsvd.n)-gsvd.k)
+	}
+	if len(s) != min(gsvd.m, gsvd.n)-gsvd.k {
+		panic(matrix.ErrSliceLengthMismatch)
+	}
+	copy(s, gsvd.s[gsvd.k:min(gsvd.m, gsvd.n)])
+	return s
+}
+
+// ZeroRFromGSVD extracts the matrix [ 0 R ] from the singular value decomposition, storing
+// the result in-place into the receiver. [ 0 R ] is size (k+l)×n.
+func (m *Dense) ZeroRFromGSVD(gsvd *GSVD) {
+	if gsvd.kind == 0 {
+		panic("gsvd: no decomposition computed")
+	}
+	m.reuseAsZeroed(gsvd.k+gsvd.l, gsvd.n)
+	a := Dense{
+		mat:     gsvd.a,
+		capRows: gsvd.m,
+		capCols: gsvd.n,
+	}
+	m.Slice(0, min(gsvd.k+gsvd.l, gsvd.m), gsvd.n-gsvd.k-gsvd.l, gsvd.n).(*Dense).
```

Comment:
I'll go through and s/m/r/g and s/n/c/g since otherwise I have collisions with the receiver name.